src/Domain/CoordinateMaps/CubicScale.cpp

```diff
@@ -18,37 +23,83 @@
 #include "Utilities/ConstantExpressions.hpp"
 #include "Utilities/DereferenceWrapper.hpp"
 #include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
 #include "Utilities/MakeWithValue.hpp"
+#include "Utilities/StdArrayHelpers.hpp"
+#include "Utilities/TypeTraits.hpp"
 
 namespace domain {
 namespace CoordMapsTimeDependent {
 
-CubicScale::CubicScale(const double outer_boundary) noexcept
-    : outer_boundary_(outer_boundary) {
+template <size_t Dim>
+CubicScale<Dim>::CubicScale(const double outer_boundary,
+                            std::string function_of_time_name_a,
+                            std::string function_of_time_name_b) noexcept
+    : f_of_t_a_(std::move(function_of_time_name_a)),
+      f_of_t_b_(std::move(function_of_time_name_b)),
+      outer_boundary_(outer_boundary),
+      functions_of_time_equal_(f_of_t_a_ == f_of_t_b_) {
   if (outer_boundary_ <= 0.0) {
     ERROR("For invertability, we require outer_boundary to be positive.\n");
   }
 }
 
+template <size_t Dim>
 template <typename T>
-std::array<tt::remove_cvref_wrap_t<T>, 1> CubicScale::operator()(
-    const std::array<T, 1>& source_coords, const double time,
+std::array<tt::remove_cvref_wrap_t<T>, Dim> CubicScale<Dim>::operator()(
+    const std::array<T, Dim>& source_coords, const double time,
     const std::unordered_map<
         std::string, std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
         map_list) const noexcept {
   const auto a_of_t = map_list.at(f_of_t_a_)->func(time)[0][0];
+
+  if (functions_of_time_equal_) {
+    // optimization for linear radial scaling
+    std::array<tt::remove_cvref_wrap_t<T>, Dim> result{};
+    for (size_t i = 0; i < Dim; ++i) {
+      gsl::at(result, i) = a_of_t * gsl::at(source_coords, i);
+    }
+    return result;
+  }
+
   const auto b_of_t = map_list.at(f_of_t_b_)->func(time)[0][0];
-  return {{source_coords[0] *
-           (a_of_t +
-            (b_of_t - a_of_t) * square(source_coords[0] / outer_boundary_))}};
+
+  tt::remove_cvref_wrap_t<T> r_squared =
+      square(dereference_wrapper(source_coords[0]));
+  for (size_t i = 1; i < Dim; ++i) {
+    r_squared += square(dereference_wrapper(gsl::at(source_coords, i)));
+  }
+  // Reuse r^2 allocation
+  r_squared = a_of_t + (b_of_t - a_of_t) / square(outer_boundary_) * r_squared;
+
+  std::array<tt::remove_cvref_wrap_t<T>, Dim> result{};
+  for (size_t i = 0; i < Dim - 1; ++i) {
+    gsl::at(result, i) = gsl::at(source_coords, i) * r_squared;
+  }
+  r_squared = source_coords[Dim - 1] * r_squared;
```

Comment:
`gsl::at` checks at runtime that the passed index is not out of range. Since `Dim - 1` is a compile-time constant, the compiler already checks that it is not out of range. This is also why clang-tidy doesn't complain about it and why we started using `gsl::at`. Does that help?

Reply:
Ah, gotcha. That makes sense, thanks!